src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs

```diff
@@ -28,12 +28,27 @@ public override async Task<SearchResult<IPackageSearchMetadata>> ContinueSearchA
             var searchToken = continuationToken as FeedSearchContinuationToken ?? throw new InvalidOperationException(Strings.Exception_InvalidContinuationToken);
             cancellationToken.ThrowIfCancellationRequested();
 
+            PackageCollectionItem[] installedFeedItems = _installedPackages.GetLatest();
+            PackageCollectionItem[] matchesInstalled = PerformLookup(installedFeedItems, searchToken);
+
             // Remove transitive packages from project references
-            IEnumerable<PackageCollectionItem> pkgsWithOrigins = _transitivePackages
-                .Where(t => t.PackageReferences.Any(x => x is ITransitivePackageReferenceContextInfo y && y.TransitiveOrigins.Any()));
+            IEnumerable<PackageCollectionItem> transitivePkgsWithOrigins = _transitivePackages
+                .Where(transitivePkg => transitivePkg
+                    .PackageReferences
+                    .All(pkgRef => pkgRef is ITransitivePackageReferenceContextInfo trPkgRef && trPkgRef.TransitiveOrigins.Any()));
+
+            // First, Search all transitive packages
+            // If we first group transitive packages by Id, then, lookup, transitive package origins could not match Installed package version
+            PackageCollectionItem[] matchesTransitive = PerformLookupCore(transitivePkgsWithOrigins, searchToken)
+                // Then, group by package ID
+                .GroupById()
+                // Select latest version with matching transitive origin within installed packages collection
+                // or fallback to latest transitive package version available if its transitive origins are not found
+                .Select(group => GetLatestApplicableTransitivePackageVersion(group, matchesInstalled))
+                .ToArray();
 
-            IPackageSearchMetadata[] installedItems = await GetMetadataForPackagesAndSortAsync(PerformLookup(_installedPackages, searchToken), searchToken.SearchFilter.IncludePrerelease, cancellationToken);
-            IPackageSearchMetadata[] transitiveItems = await GetMetadataForPackagesAndSortAsync(PerformLookup(pkgsWithOrigins, searchToken), searchToken.SearchFilter.IncludePrerelease, cancellationToken);
+            IPackageSearchMetadata[] installedItems = await GetMetadataForPackagesAndSortAsync(matchesInstalled, searchToken.SearchFilter.IncludePrerelease, cancellationToken);
+            IPackageSearchMetadata[] transitiveItems = await GetMetadataForPackagesAndSortAsync(matchesTransitive, searchToken.SearchFilter.IncludePrerelease, cancellationToken);
```

Comment:
Do we have metadata cache layer here? This `GetMetadataForPackagesAndSortAsync` making http call go get metadata, but if we get metadata for hundreds/thousands of packages then delay would be significant.

Reply:
If it takes much effort then don't worry about it, I'm working on separate `http cache` effort.